vendor/github.com/xtaci/smux/session.go

```diff
@@ -281,73 +277,88 @@ func (s *Session) recvLoop() {
 	}
 }
 
-func (s *Session) keepalive() {
-	tickerPing := time.NewTicker(s.config.KeepAliveInterval)
-	tickerTimeout := time.NewTicker(s.config.KeepAliveTimeout)
-	defer tickerPing.Stop()
-	defer tickerTimeout.Stop()
+// keepAliveSend will periodically send a keepalive meesage to the remote peer.
+func (s *Session) keepAliveSend() {
+	keepAliveTimeout := time.After(s.config.KeepAliveInterval)
 	for {
 		select {
-		case <-tickerPing.C:
-			s.writeFrame(newFrame(cmdNOP, 0))
-			s.notifyBucket() // force a signal to the recvLoop
-		case <-tickerTimeout.C:
-			if !atomic.CompareAndSwapInt32(&s.dataReady, 1, 0) {
-				s.Close()
-				return
-			}
 		case <-s.die:
 			return
+		case <-keepAliveTimeout:
+			keepAliveTimeout = time.After(s.config.KeepAliveInterval) // set before writing so we start sending the next one in time
```

Comment:
if the timeout channel is reset immediately upon reading from the channel... what's the difference between this and a ticker?

I looked into the `time.Ticker` source code, and it appears that its channel has a buffer of 1. So if you're busy doing some other work and you "miss" a tick, the channel will still have a value in it when you come back, and thus the next read will succeed immediately. If you miss multiple ticks, `time.Ticker` skips the send (leaving the channel with one value in it). I think this prevents the "missed tick" behavior that your comment describes.

So if we did:
```go
sendTicker := time.NewTicker(s.config.KeepAliveInterval)
defer sendTicker.Stop()
for {
	select {
	case <-s.die:
		return
	case <-sendTicker.C:
		writeFrame()
	}
}
```
I think it's functionally equivalent. We can check by examining three scenarios:
1. `writeFrame` always takes less time than `s.config.KeepAliveInterval`
2. `writeFrame` always takes more time than `s.config.KeepAliveInterval`
3. `writeFrame` varies between taking more time and less time than `s.config.KeepAliveInterval`.

We'll use a `KeepAliveInterval` of 1 second for each case. In the first case, `writeFrame` will take 0.6 seconds. In the second, it will take 1.4 seconds. In the last case, it will alternate between 0.7 seconds and 1.5 seconds.

Case 1:

In case 1, both the PR code and the Ticker code call `writeFrame` at `t=1`, `t=2`, `t=3`, etc. `writeFrame` never affects the timing.

Case 2:

In case 2, both the PR code and the Ticker code call `writeFrame` at `t=1`, `t=2.4`, `t=3.8`, etc. Since the channel is filled during the `writeFrame` call, the next channel read always succeeds immediately.

Case 3:

In case 3, the PR code calls `writeFrame` at `t=1`, `t=2`, `t=3.5`, `t=4.5`, `t=6`, `t=7`... whereas the Ticker code calls `writeFrame` at `t=1`, `t=2`, `t=3.5`, `t=4.2`, `t=5.7`, `t=6.4`...

So these implementations are in fact different! Here's why. In the PR code, when `writeFrame` is called at `t=3.5`, it takes 0.7 seconds, finishing at `t=4.2`. Since `time.After` was called immediately prior to `writeFrame`, it sends down the channel at `t=4.5`, which kicks off the next `writeFrame`.

In the Ticker code, `writeFrame` is again called at `t=3.5` and finishes at `t=4.2`. The Ticker is dutifully ticking on each second, so during the `writeFrame` call, it sends down the channel at `t=4`. When `writeFrame` finishes at `t=4.2`, the channel is full, so the next `writeFrame` is called immediately.

[Playground link](https://play.golang.org/p/5pVAumYGzRe)

I'm not sure which version is preferable. Both seem to call `writeFrame` at reasonable intervals. Ideally we could devise a test cases that causes unacceptable behavior in one of them, but not the other.